File: mrbgems/mruby-signal/src/signal.c
/*
** signal.c - Signal module
**
** See Copyright Notice in mruby.h
**
** The common half of mruby-signal: the two questions Ruby asks about a
** signal name, answered over the platform table in signal_hal.h.  Which
** names a host has is the port's business; that `EXIT` is one of them, and
** that a number naming no signal answers nil rather than raising, is Ruby's
** and is settled here.
*/

#include <mruby.h>
#include <mruby/hash.h>
#include <mruby/string.h>
#include "signal_hal.h"

/*
 * Signal 0, and the name Ruby gives it.
 *
 * No platform has a signal numbered 0: `kill(2)` spends it on asking whether
 * a process can be signalled at all, and Ruby spends the name on the handler
 * that runs at exit.  It is a name Ruby adds rather than one a host reports,
 * so it is added here, where every port gets it alike.
 */
#define SIGNAL_EXIT_NAME "EXIT"
#define SIGNAL_EXIT_NUMBER 0

/*
 * call-seq:
 *   Signal.signame(signo) -> string or nil
 *
 * The bare name this platform gives signal +signo+, without the "SIG"
 * prefix, or nil where the number names no signal here.
 *
 *   Signal.signame(9)   # => "KILL"
 *   Signal.signame(0)   # => "EXIT"
 *   Signal.signame(999) # => nil
 *
 * Where a host spells one signal two ways, the name Ruby lists first is the
 * one that comes back: signal 6 is "ABRT" rather than "IOT".
 */
static mrb_value
signal_s_signame(mrb_state *mrb, mrb_value self)
{
  mrb_int signo;
  const char *name;

  mrb_get_args(mrb, "i", &signo);
  if (signo == SIGNAL_EXIT_NUMBER) {
    return mrb_str_new_lit(mrb, SIGNAL_EXIT_NAME);
  }
  /* Nothing is narrowed on the way to the table, where the comparison is in
     mrb_int, so a number too large for a signal simply matches no entry and
     comes back nil, as it does in CRuby. */
  name = mrb_hal_signal_name(mrb, signo);
  if (name == NULL) return mrb_nil_value();
  return mrb_str_new_cstr(mrb, name);
}

/*
 * call-seq:
 *   Signal.list -> hash
 *
 * Every signal name this platform knows, mapped to its number.  A signal a
 * host spells two ways appears under each spelling:
 *
 *   Signal.list["ABRT"] # => 6
 *   Signal.list["IOT"]  # => 6
 *
 * The hash is built fresh each call, so changing it changes nothing.
 */
static mrb_value
signal_s_list(mrb_state *mrb, mrb_value self)
{
  mrb_int count = mrb_hal_signal_count(mrb);
  mrb_value list = mrb_hash_new_capa(mrb, count + 1);
  int ai = mrb_gc_arena_save(mrb);
  mrb_int i;

  mrb_hash_set(mrb, list, mrb_str_new_lit(mrb, SIGNAL_EXIT_NAME),
               mrb_int_value(mrb, SIGNAL_EXIT_NUMBER));
  mrb_gc_arena_restore(mrb, ai);

  for (i = 0; i < count; i++) {
    mrb_int signo;
    const char *name = mrb_hal_signal_at(mrb, i, &signo);

    if (name == NULL) break;
    mrb_hash_set(mrb, list, mrb_str_new_cstr(mrb, name), mrb_int_value(mrb, signo));
    mrb_gc_arena_restore(mrb, ai);
  }
  return list;
}

void
mrb_mruby_signal_gem_init(mrb_state *mrb)
{
  struct RClass *signal = mrb_define_module_id(mrb, MRB_SYM(Signal));

  mrb_define_module_function_id(mrb, signal, MRB_SYM(signame), signal_s_signame, MRB_ARGS_REQ(1));
  mrb_define_module_function_id(mrb, signal, MRB_SYM(list),    signal_s_list,    MRB_ARGS_NONE());
}

void
mrb_mruby_signal_gem_final(mrb_state *mrb)
{
}

File: mrbgems/mruby-signal/test/signal.rb
##
# Signal ISO Test

assert('Signal.signame') do
  assert_equal "KILL", Signal.signame(9)
  assert_equal "TERM", Signal.signame(15)
  assert_equal "INT", Signal.signame(2)
end

assert('Signal.signame with a number that names no signal') do
  # Nothing here is an error: Ruby answers nil for a number the platform does
  # not name, however far out it is.
  assert_nil Signal.signame(-1)
  assert_nil Signal.signame(999)
end

assert('Signal.signame(0) is EXIT') do
  # No platform numbers a signal 0, so the name is Ruby's own rather than one
  # a port reports, and every platform answers with it.
  assert_equal "EXIT", Signal.signame(0)
end

assert('Signal.signame refuses what is not a number') do
  assert_raise(TypeError) { Signal.signame("9") }
  assert_raise(TypeError) { Signal.signame(nil) }
  assert_raise(TypeError) { Signal.signame(:KILL) }
end

assert('Signal.signame answers with the name Ruby answers with') do
  # A host that spells one signal two ways gives both names the same number,
  # and the table is ordered so that the reverse lookup finds the name Ruby
  # reports: ABRT rather than IOT, CHLD rather than CLD, IO rather than POLL.
  aliases = %w[IOT CLD POLL]
  0.upto(64) do |signo|
    name = Signal.signame(signo)
    assert_not_include aliases, name if name
  end
end

assert('Signal.list') do
  list = Signal.list
  assert_kind_of Hash, list
  assert_equal 0, list["EXIT"]
  assert_equal 9, list["KILL"]
  assert_equal 15, list["TERM"]
  assert_nil list["SIGKILL"]  # the names are bare, without the prefix
end

assert('Signal.list agrees with Signal.signame') do
  # Every name maps to a number that names a signal, and the name that comes
  # back maps to the same number.  An alias fails the first test and passes
  # the second, which is what makes the two directions consistent rather than
  # merely reversible.
  Signal.list.each do |name, signo|
    assert_kind_of String, name
    assert_kind_of Integer, signo
    back = Signal.signame(signo)
    assert_not_nil back
    assert_equal signo, Signal.list[back]
  end
end

assert('Signal.list is built fresh each call') do
  list = Signal.list
  list["NOSUCHSIGNAL"] = 1234
  assert_nil Signal.list["NOSUCHSIGNAL"]
end
